loki/transformations/temporaries/stack_allocator.py

```diff
@@ -0,0 +1,1372 @@
+            The items corresponding to successor routines called from :data:`routine`
+        """
+        successor_map = {
+            successor.local_name: successor
+            for successor in successors if isinstance(successor, ProcedureItem)
+        }
+        call_map = {}
+
+        # loop over calls and check if they call a successor routine and if the
+        # transformation data is available
+        for call in FindNodes(CallStatement).visit(routine.body):
+            if call.name in successor_map and self._key in successor_map[call.name].trafo_data:
+                successor_stack_dict = successor_map[call.name].trafo_data[self._key]['stack_dict']
+
+                call_stack_args = []
+
+                # loop over dtypes and kinds in successor arguments stacks
+                # and construct list of stack arguments
+                for dtype in successor_stack_dict:
+                    for kind in successor_stack_dict[dtype]:
+                        call_stack_args += list(stack_arg_dict[dtype][kind])
+
+                # get position of optional arguments so we can place the stacks in front
+                arg_pos = [call.routine.arguments.index(arg) for arg in call.routine.arguments if arg.type.optional]
+
+                arguments = call.arguments
+                if arg_pos:
+                    # stack arguments have already been added to the routine call signature
+                    # so we have to subtract the number of stack arguments from the optional position
+                    arg_pos = min(arg_pos) - len(call_stack_args)
+                    arguments = arguments[:arg_pos] + as_tuple(call_stack_args) + arguments[arg_pos:]
+                else:
+                    arguments += as_tuple(call_stack_args)
+
+                call_map[call] = call.clone(arguments=arguments)
+
+        if call_map:
+            routine.body = Transformer(call_map).visit(routine.body)
+
+
+    def create_stacks_driver(self, routine, stack_dict, successors):
+        """
+        Create stack variables in the driver routine,
+        add pragma directives to create the stacks on the device,
+        and add the stack_variables to kernel call arguments.
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            The driver subroutine to get the stack_variables
+        stack_dict : dict
+            dict that maps dtype and kind to an expression for the required stack size
+        successors : list of :any:`Item`
+            The items corresponding to successor routines called from :data:`routine`
+        """
+
+        # block variables
+        kgpblock = self._get_int_var(name=self.block_dim.size, scope=routine)
+        jgpblock = self._get_int_var(name=self.block_dim.index, scope=routine)
+
+        stack_vars = []
+        stack_arg_dict = {}
+        assignments = []
+        deallocs = []
+        pragma_string = ''
+        pragma_data_start = None
```

Comment:
This is used only locally further down in the conditional, so shouldn't need to be initialized here
```suggestion
```